public/pages/CreateDetector/components/DefineDetector/components/DetectionRules/DetectionRules.tsx

```diff
@@ -0,0 +1,186 @@
+    const ruleTypes = Object.keys(rulesByRuleType);
+
+    return (
+      <EuiPanel style={{ paddingLeft: '0px', paddingRight: '0px' }}>
+        <EuiAccordion
+          buttonContent={
+            <EuiTitle>
+              <h4>{`Threat detection rules (${allRulesCount})`}</h4>
+            </EuiTitle>
+          }
+          buttonProps={{ style: { paddingLeft: '10px', paddingRight: '10px' } }}
+          id={'detectorRulesAccordion'}
+          initialIsOpen={false}
+        >
+          <EuiHorizontalRule margin={'xs'} />
+          <div style={{ paddingLeft: '10px', paddingRight: '10px' }}>
+            <EuiSpacer size={'m'} />
+            <EuiFlexGroup>
+              <EuiFlexItem grow={false}>
+                <EuiText>
+                  <EuiButtonEmpty onClick={() => this.onRuleTypeClick()}>
+                    View all rules {`(${allRulesCount})`}
+                  </EuiButtonEmpty>
+                  <EuiSpacer />
+                  {ruleTypes.map((ruleType) => (
+                    <React.Fragment key={ruleType}>
+                      <EuiButtonEmpty
+                        onClick={() => this.onRuleTypeClick(ruleType)}
+                      >{`${ruleType} (${rulesByRuleType[ruleType].ruleItems.length})`}</EuiButtonEmpty>
+                      <EuiSpacer size="xs" />
+                    </React.Fragment>
+                  ))}
+                </EuiText>
+              </EuiFlexItem>
+              <EuiFlexItem>
+                <ContentPanel
+                  title={`${
+                    selectedRuleType || 'All'
+                  } rules (${activeRulesCountForSelectedType}/${totalRulesCountForSelectedType} enabled)`}
+                >
+                  <EuiBasicTable
+                    columns={getRulesColumns((changedItem: RuleItem, changeToActive: boolean) => {
+                      const ruleItems = rulesByRuleType[changedItem.ruleType].ruleItems;
+                      const changedIdx = ruleItems.findIndex(
+                        (item) => item.ruleName === changedItem.ruleName
+                      );
+
+                      if (changedIdx > -1) {
+                        const newRuleItems = [
+                          ...ruleItems.slice(0, changedIdx),
+                          { ...ruleItems[changedIdx], active: changeToActive },
+                          ...ruleItems.slice(changedIdx + 1),
+                        ];
+                        const newRulesByRuleType: RulesInfoByType = {
+                          ...rulesByRuleType,
+                          [changedItem.ruleType]: {
+                            ruleItems: newRuleItems,
+                            activeCount:
+                              rulesByRuleType[changedItem.ruleType].activeCount +
+                              (changeToActive ? 1 : -1),
+                          },
+                        };
+                        this.setState({ rulesByRuleType: newRulesByRuleType });
+                      }
```

Comment:
I would suggest keeping lines 151-174 out of your JSX, maybe put it in the render(), especially setState seems like it could lead to issues being there.  What do you think?

Reply:
yeah, makes sense. Let me extract it out. The setState won't get called as part of render but instead of change in a field, will it make it more apparent.

Reply:
extracted the logic into a method